cpp/include/ray/api/ray_runtime_holder.h

```diff
@@ -0,0 +1,45 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <ray/api/ray_runtime.h>
+
+namespace ray {
+namespace api {
+
+struct RayRuntimeHolder {
+  static RayRuntimeHolder &Instance() {
+    static RayRuntimeHolder instance;
+    return instance;
+  }
+
+  void Init(std::shared_ptr<RayRuntime> runtime) { runtime_ = runtime; }
+
+  std::shared_ptr<RayRuntime> &Runtime() { return runtime_; }
+
+ private:
+  RayRuntimeHolder() = default;
+  ~RayRuntimeHolder() = default;
+  RayRuntimeHolder(RayRuntimeHolder const &) = delete;
+  RayRuntimeHolder(RayRuntimeHolder &&) = delete;
+  RayRuntimeHolder &operator=(RayRuntimeHolder const &) = delete;
+  RayRuntimeHolder &operator=(RayRuntimeHolder &&) = delete;
+
+  std::shared_ptr<RayRuntime> runtime_;
+};
+
+static auto &ray_runtime = RayRuntimeHolder::Instance().Runtime();
```

Comment:
```suggestion
static auto &ray_runtime_instance = RayRuntimeHolder::Instance().Runtime();
```

Reply:
It seems that this is more like a function, not a variable. What if this:

```
Ray::Init();
Ray::Shutdown();
Ray::Init();
```

Will `ray_runtime` automatically points to the second runtime?

Reply:
As i mentioned before: https://github.com/ray-project/ray/pull/14414#discussion_r585281434, we need to move it in a independent class to avoid circular references.

RayRuntimeHolder::Instance().Runtime() return a reference of RayRuntimeHolder::runtime_, so the ray_runtime is a reference of RayRuntimeHolder::runtime_.

Reply:
I mean the static variable always points to the first runtime instance, even if the first runtime instance is destructed and the second runtime instance is created, right? In this case, accessing the static variable may results in SIGSEGV.

Reply:
I think it's ok, because we use std::call_once to make sure the ray_runtime init only once, however maybe use a function to return it more safer.

Reply:
So do you mean that calling `Ray::Init()` and `Ray::Shutdown()` multiple times is not supported in C++? They are supported in both Python and Java.

Reply:
We can call Ray::Init many times, in the Init function we use make sure init only once, avoid duplicate init.
It's same with Python and Java.

Reply:
I mean

```
Ray::Init();
Ray::Shutdown();
Ray::Init();
```

not

```
Ray::Init();
Ray::Init();
```

Reply:
I did a test, it works but the the second internal init is not triggered, i think it's a bug, it should be init again as we have shutdown ray.

I think we can fix it in another PR.

Reply:
Sure.